packages/fx-core/src/core/middleware/projectVersionChecker.ts

```diff
@@ -51,34 +54,33 @@ async function needToShowUpdateDialog(ctx: CoreHookContext, versionInfo: Version
 }
 
 // TODO: add url for download proper toolkit version
-async function showDialog(ctx: CoreHookContext) {
+async function showDialog(ctx: CoreHookContext): Promise<FxError> {
   const lastArg = ctx.arguments[ctx.arguments.length - 1];
   const inputs: Inputs = lastArg === ctx ? ctx.arguments[ctx.arguments.length - 2] : lastArg;
   if (inputs.platform === Platform.VSCode) {
-    const res = await TOOLS?.ui.showMessage(
-      "warn",
-      getLocalizedString("core.projectVersionChecker.incompatibleProject"),
-      false,
-      learnMoreText
-    );
-    if (res.isOk() && res.value === learnMoreText) {
-      TOOLS?.ui!.openUrl(MetadataV2.updateToolkitLink);
-    }
+    const messageKey = "core.projectVersionChecker.incompatibleProject";
+    const message = getLocalizedString(messageKey);
+    TOOLS?.ui.showMessage("warn", message, false, learnMoreText)?.then((res) => {
+      if (res.isOk() && res.value === learnMoreText) {
+        TOOLS?.ui!.openUrl(MetadataV2.updateToolkitLink);
+      }
+    });
+    return IncompatibleProjectError(messageKey);
   } else if (inputs.platform === Platform.CLI) {
-    TOOLS?.logProvider.warning(getLocalizedString("core.projectVersionChecker.cliUseNewVersion"));
-  } else if (inputs.platform === Platform.VS) {
-    const res = await TOOLS?.ui.showMessage(
-      "warn",
-      getLocalizedString(
-        "core.projectVersionChecker.vscodeUseNewVersion",
-        "Visual Studio 2022 17.5 Preview"
-      ),
-      false,
-      learnMoreText
-    );
-    if (res.isOk() && res.value === learnMoreText) {
-      TOOLS?.ui!.openUrl(learnMoreLink);
-    }
+    const messageKey = "core.projectVersionChecker.cliUseNewVersion";
+    TOOLS?.logProvider.warning(getLocalizedString(messageKey));
+    return IncompatibleProjectError(messageKey);
+  } else {
+    const messageKey = "core.projectVersionChecker.incompatibleProject";
+    const expectedVersion = "Visual Studio 2022 17.5 Preview";
```

Comment:
why we suggest a version for VS but not VS Code? By the way, `Visual Studio 2022 17.5 Preview` should no longer be suggested because some things changed since its release.

Reply:
As this fix is main branch, it works for cli and will not affect vs release.
vs message has already updated in dev branch, it will not contain detailed version information to align with vscode/cli messages.